### PR TITLE
fix torch.tensor for functionalization

### PR DESCRIPTION
**Pull request by bdhirsh**
Right now, using the `torch.Tensor` constructor inside of a functionalized function is broken (and there's a request to use it during tracing for mobile: https://fb.workplace.com/groups/1405155842844877/permalink/5805679106125840/)

update: I took the approach described in the comments (letting `at::empty()` run directly with the data, and using `.to()` to "lift" it into a wrapper), which required a minor change described below.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #77132
* #77131
* #77130
* #77129
* #77128
* #77127
* #77126
* #77125
* #76320
* #76126
* __->__ #76319

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:296]: It occurs to me that this actually is very inefficient, right?! We allocate a sizes storage, and then throw it out immediately after! If we had some sort of `new_as_strided` (which @albanD was mumbling about at https://github.com/pytorch/pytorch/pull/75994 ) we could do this all in one go, it would be faster, and you could directly implement functionalization there.

Comment (albanD) [reply torch/csrc/utils/tensor_new.cpp:296]: We do have the low level function in c++: https://github.com/pytorch/pytorch/blob/1d55518198de14a011e7acd6c604afe512ee28b4/aten/src/ATen/native/native_functions.yaml#L1902
The new version could be added!

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:312]: This is a long comment saying why the thing doesn't work, but I think what I'd actually read about is how, morally it should work

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:316]: how come `data_tensor` gets set in one branch but not the other?

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:332]: ...shadowing the data_tensor above?

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:314]: This is the "storage was passed to tensor constructor" code path, doesn't anyone actually *need* this?

If we want to do this soundly we need to identify if the passed in storage is a functionalization storage or a regular storage, because it seems to me like unwrapping the functional tensor if its a functionalization storage would be the wrong thing to do.

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:339]: It should be sound to compute this on tensor.sizes() too right?

**Comment by ezyang** [comment]
The code at stake here is not big, so I don't think it's too risky to land this as is (esp as an unblocker). However, I think I have an alternative suggestion for how to do this properly. The general concept of what is going on here is that `torch.tensor` is all about creating a "raw" tensor (directly writing in data) without any transforms in effect, and then "lifting/wrapping" it into a constant in the relevant transform. So what I think should happen is that the "data" manipulations should happen under some RAII guard that *disables* all transforms, and then we need a way to lift/wrap based on all of the currently active transforms that require mandatory wrapping immediately.

**Comment by bdhirsh** [comment]
> So what I think should happen is that the "data" manipulations should happen under some RAII guard that disables all transforms, and then we need a way to lift/wrap based on all of the currently active transforms that require mandatory wrapping immediately.

Hmm ok, I originally wasn't sure if this was the end-state behavior that we wanted (partially because of the weird interaction that it would cause with LTC/XLA), but the description makes sense to me. I'll switch it over to do it this way in the PR. We can also get things to work out on the LTC/XLA side by having their `empty` kernel explicitly read out the TLS to know whether or not to do the wrapping.

**Comment by ezyang** [comment]
I didn't think carefully about the LTC/XLA side of things. But it seems similar? You need to make an honest to goodness tensor with the data you want, and then lower it into the XLA graph as a constant. That's what lift should be doing, I think?

File: aten/src/ATen/FunctionalizeFallbackKernel.cpp
Comment (bdhirsh) [inline aten/src/ATen/FunctionalizeFallbackKernel.cpp:102]: I updated the PR to use the approach where we construct an ordinary empty tensor and "lift" into a wrapper with the `.to()` call.

That required some changes though, since `to()` is `CompositeImplicitAutograd` (and by default functionalization will decompose all composite ops). If the `.to()` call no-ops (which it will in most cases, if the output dtype + device match), then we would just never hit a functionalization kernel and do the lifting.

I also can't use the boxed fallback (needed to write a custom kernel), since in the normal case, we *want* `.to()` to decompose. `to()` is actually a "problematic" op for functionalization, since it may-or-may-not return a view depending on whether or not it no-ops.

Comment (ezyang) [reply aten/src/ATen/FunctionalizeFallbackKernel.cpp:102]: `.to()` seems like the wrong way to get to a functionalize tensor...

Comment (ezyang) [reply aten/src/ATen/FunctionalizeFallbackKernel.cpp:102]: It seems like the constraint that you posed for yourself was to not have to modify the tensor_new.cpp (besides excluding functionalize key) but I think that is too much of a handicap. Once the tensor has been moved to the correct device, I think you should be willing to call another *new* operator that does the lifting.

Comment (bdhirsh) [reply aten/src/ATen/FunctionalizeFallbackKernel.cpp:102]: That sounds reasonable - I have two more questions:

(1) Should the other stuff in this function (functorch, autograd, tracer key, python key) also be updated to do the "lifting" separately from the `.to()` call? (Well I don't think it really makes sense for autograd, and it would be BC-breaking for tracing to trace a new op. Technically BC-breaking for python too, although maybe that's a better end state?)

(2) Should the "lifting" be an aten operator? (I think the answer is yes if we want to be able to do the lifting differently for different transforms, although lifting doesn't really make sense for e.g. backends).

Comment (ezyang) [reply aten/src/ATen/FunctionalizeFallbackKernel.cpp:102]: > Should the other stuff in this function (functorch, autograd, tracer key, python key) also be updated to do the "lifting" separately from the .to() call?

What makes this confusing is that traditionally, lifting is a no-op (and it should continue to be a no-op for all the cases it can be). The second confusing thing is that lifting has to be relative to *something*. In this case, we lift to the ambient mode (if there was no mode, lifting wouldn't do anything--but you could also imagine a variant of lifting that lifts to match another tensor; see for example https://github.com/albanD/subclass_zoo/blob/a4a0052832edaf6fb538095894c6c61ca6d5ba6a/simple_functorch.py#L896-L911 )

But yes, I think it should be an aten operator, but it should do nothing most of the time.

**Comment by zhxchen17** [comment]
@bdhirsh any updates on this diff? I'm wondering are we ready to land this? I'm not blocked by this, but just curious to know our progress here.

**Review by ezyang — APPROVED** [review]

**Comment by bdhirsh** [comment]
@zhxchen17 I've been working on a big stack of functionalization changes locally including the feedback from this PR. Actually just pushed out the changes a minute ago.

@ezyang I saw that you just gave the PR an approve, feel free to take a look at the new changes - I added a new `lift` aten op, that currently only has an implementation for functionalization (and has a default "no-op" implementation).

I'm going to need to add a companion PR for functorch before I can land this though - the existing wrapper subclasses in functorch need to know about "lifting"

File: torch/csrc/utils/tensor_new.cpp
Comment (ezyang) [inline torch/csrc/utils/tensor_new.cpp:285]: It feels like there should be an easy way to just "exclude everything", but we can probably work that out later.

**Review by ezyang — APPROVED** [review]
I reviewed the new code with lift and I like it a lot! Thanks!

**Comment by bdhirsh** [comment]
I'm actually pretty sure that this doesn't require functorch changes: I was trying to change functorch locally, and my understanding is that now the code in the `FuncTorchDynamicLayerFrontMode` fallback kernel is responsible for the wrapping. Since that key is still enabled during the `.to()` call, I think the wrapping still happens correctly without functorch adding support for `lift()`.

Getting functorch to support `lift()` seems like the right the to do but I'd rather land this PR now to unblock mobile.

It also looks like functorch CI is failing on master, but I was able to run `python test/test_eager_transforms.py TestGradTransformCPU` locally without any failures (which includes tests that run the `torch.tensor()` constructor under the grad transform).

**Pull request closed without merging**
